### DISCUSSION TITLE
Differences in BLAST-ings

### DISCUSSION THREAD

**Discussion by kubu4** [RobertsLab/resources#1691] [Q&A]
I'll start out by qualifying this post as possibly being moot from the start, as the following is definitely not an "apples to apples" comparison. Okay, with that out of the way...

Recently, Steven and I took two different approaches to annotating (via BLASTing) the _C.virginica_ genome and we got very different results in the number of BLAST matches.

#### @sr320's approach ([Notebook](https://sr320.github.io/annotation/))

1. Use translated CDS FastA as queries ([GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/022/765/GCF_002022765.2_C_virginica-3.0/GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz))
    - 60,213 sequences (`grep -c "^>" GCF_002022765.2_C_virginica-3.0_translated_cds.faa`)
    - 34,608 unique genes (`grep -o '\[gene=[^]]*\]' | grep -oP '(?<=\[gene=)[^]]*' GCF_002022765.2_C_virginica-3.0_translated_cds.faa | sort -u | wc -l`)
2. Use reviewed UniProt proteins FastA as BLAST database ([uniprot_sprot.fasta.gz](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz)).
3. Use NCBI BLAST to make BLAST database.
4. Use NCBI BLASTp to BLAST.
5. Do some gene ontology stuff.

Results:

#### BLAST
- 46,130 matches (`wc -l Cvir_transcds-uniprot_blastp.tab`)
- 10,370 _unique_ SwissProt IDs (`awk -F "|" '{print $3}' Cvir_transcds-uniprot_blastp.tab | sort -u | wc -l`)

#### Gene Ontology
- 17,971 genes annotated

---

#### Sam's approach

1. [Extract gene sequences as FastA](https://robertslab.github.io/sams-notebook/2023/07/26/Data-Wrangling-C.virginica-Genes-Only-FastA-from-Genes-BED-File-Using-gffread-on-Raven.html) (Notebook).

    - 38,838 genes (`wc -l C_virginica-3.0_Gnomon_genes.bed`)
2. Use reviewed UniProt proteins FastA as BLAST database ([uniprot_sprot.fasta.gz](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz)).
3. Use DIAMOND to make DIAMOND BLAST database.
4. [Use DIAMOND BLASTx to BLAST](https://robertslab.github.io/sams-notebook/2023/07/27/DIAMOND-BLASTx-C.virginica-Genes-on-Mox.html). (Notebook)

Results:

#### BLAST
- 5,869 matches (`wc -l GCF_002022765.2_C_virginica-3.0-genes.blastx.outfmt6`)
- 3,113 _unique_ SwissProt IDs (`awk -F "|" '{print $2}' GCF_002022765.2_C_virginica-3.0-genes.blastx.outfmt6 | sort -u | wc -l`)

---

Sooo, what do people think about this? Anyone have any thoughts/opinions? E-value cutoffs for BLASTing for both approaches were 1e<sup>-20</sup> and 1e<sup>-25</sup> for each approach, respectively.

**Comment by sr320** [comment]
One thing I would be interested in is head to head comparison of blast and diamond blast (same query/db/sertings)

**Reply by kubu4** [reply]
I ran an NCBI BLASTx, which is comparable to DIAMOND BLASTx ([output directory with files](https://gannet.fish.washington.edu/Atumefaciens/20230801-cvir-ncbi-blastx-GCF_002022765.2_C_virginica-3.0-genes/)).

I say "comparable" because I used the same e-value cutoff and the same UniProt protein database (i.e. constructed from the same starting FastA). However, DIAMOND BLASTx has some ["sensitivity" settings that can be set](https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options) and I'm not sure how those compare to NCBI BLASTx...

Anyway, here is a table of the two results:

|                  | DIAMOND  | NCBI     |
|------------------|----------|----------|
| Matches          | 5,869    | 6,025   |
| Unique SwissProt | 3,113    | 3,029    |
| Run time         | 00:01:30 | 17:07:02 |

The difference in run times is wild! Haha! 

DIAMOND: 1 _minute_ and 30secs
NCBI: 17hrs

NOTE: The NCBI BLASTx had multiple hits per query, despite using `-num_target_seqs 1` set... The number of matches reported above are _unique_ matches.

**Reply by sr320** [reply]
Very interesting! Note I think '-max_hsps 1' added to max target will get you uniques.

**Reply by kubu4** [reply]
> Note I think '-max_hsps 1' added to max target will get you uniques

Thanks. I haven't used NCBI BLAST in so long that I couldn't remember the magic combination to get unique matches.

**Comment by kubu4** [comment]
Personally, I'm happy to see similar results between the two when using BLASTx, but it doesn't clear up why there's such a large difference between the BLASTp and BLASTx results - which is what prompted this discussion. I still find that difference surprising.
